game/scripts/vscripts/abilities/oaa_arcane_orb.lua

```diff
@@ -31,175 +43,209 @@ function modifier_oaa_arcane_orb:RemoveOnDeath()
   return false
 end
 
-function modifier_oaa_arcane_orb:OnCreated()
-  if IsServer() then
-    if not self.procRecords then
-      self.procRecords = {}
-    end
-    self.parentOriginalProjectile = self:GetParent():GetRangedProjectileName()
-    Debug.EnabledModules["abilities:oaa_arcane_orb"] = false
-  end
-end
-
-modifier_oaa_arcane_orb.OnRefresh = modifier_oaa_arcane_orb.OnCreated
-
 function modifier_oaa_arcane_orb:DeclareFunctions()
   return {
     MODIFIER_EVENT_ON_ATTACK_START,
     MODIFIER_EVENT_ON_ATTACK,
     MODIFIER_EVENT_ON_ATTACK_LANDED,
-    MODIFIER_EVENT_ON_ATTACK_FAIL
   }
 end
 
-function modifier_oaa_arcane_orb:OnAttackStart(keys)
+function modifier_oaa_arcane_orb:OnAttack(event)
   local parent = self:GetParent()
+  local ability = self:GetAbility()
 
-  if keys.attacker ~= parent then
+  if event.attacker ~= parent then
     return
   end
 
-  local ability = self:GetAbility()
-  local target = keys.target
-  -- Wrap in function to defer evaluation
-  local function autocast()
-    return (
-      target.GetUnitName and -- Check for existence of GetUnitName method to determine if target is a unit
-      ability:GetAutoCastState() and
-      not parent:IsSilenced() and
-      ability:IsOwnersManaEnough() and
-      ability:IsOwnersGoldEnough(parent:GetPlayerOwnerID()) and
-      ability:IsCooldownReady() and
-      ability:CastFilterResultTarget(target) == UF_SUCCESS
-    )
+  if parent:GetCurrentActiveAbility() ~= ability then
+    return
   end
 
-  if parent:GetCurrentActiveAbility() ~= ability and not autocast() then
+  -- To prevent crashes:
+  local target
+  if event.target == nil then
     return
+  else
+    target = event.target
   end
 
-  -- Add modifier to change attack sound
-  parent:AddNewModifier(parent, ability, "modifier_oaa_arcane_orb_sound", {})
-  -- Set projectile
-  parent:ChangeAttackProjectile()
+  if target:IsNull() then
+    return
+  end
+
+  -- This happens only when Arcane Orb is cast manually.
+  self.manual_cast = true
+
+  if ability:IsOwnersManaEnough() and ability:IsCooldownReady() and (not parent:IsSilenced()) and (not target:IsMagicImmune()) then
+    parent:SetRangedProjectileName("particles/units/heroes/hero_obsidian_destroyer/obsidian_destroyer_arcane_orb.vpcf")
```

Comment:
I think OnAttack was supposed to change the projectile back to normal. Cause this event fires just after the projectile already flies.

Reply:
That's really late though. I think OnAttackFinished might work?

Reply:
Why make it complicated? Its not that late.

Reply:
It can be really late though (think TPs), and I think in the case of misses or disjointed attacks, it'll just never even fire.

Reply:
Yeah you are right. OnAttackFinished will work.